internal/common/metrics/handler.go

```diff
@@ -0,0 +1,138 @@
+// The MIT License
+//
+// Copyright (c) 2021 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import "time"
+
+// Handler is a handler for metrics emitted by the SDK. This interface is
+// intentionally limited to only what the SDK needs to emit metrics and is not
+// built to be a general purpose metrics abstraction for all uses.
+//
+// A common implementation is at
+// go.temporal.io/sdk/contrib/tally.NewMetricsHandler. The NopHandler is an
+// noop handler.
+type Handler interface {
```

Comment:
I think we should give user an option to provide histogram metrics as well.

Reply:
> unless user casts Handler to implementation and exposes functionality there.

Even this isn't safe because we wrap the handler in a "replay aware handler".

I have now exposed the scope from the Tally side if they want to use histograms. After internal discussion, since there isn't yet a good abstraction for histograms for OpenTelemetry and Tally, we are just exposing this value and asking them to be careful of replay. If a good histogram abstraction can be made, we can add it.